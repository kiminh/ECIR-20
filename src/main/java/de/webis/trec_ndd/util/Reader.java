package de.webis.trec_ndd.util;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

import lombok.SneakyThrows;

public class Reader<T> implements Iterator<T>, Iterable<T>{
	
	private Class<T> readerType;
	private BufferedReader[] readerList;
	private BufferedReader currentReader;
	private int currentReaderIndex=0;
	private String lastLine=null;
	private File fileToBeRead;
	
	private Reader(Reader<T> r) throws IOException{
		init(r.fileToBeRead, r.readerType);
	}

	
	public Reader(File file, Class<T> cls) throws IOException{
		init(file,cls);
	}
	
	public Reader(String sfile, Class<T> cls) throws IOException{
		init(new File(sfile),cls);
	}
	
	public void init(File file, Class<T> cls) throws IOException{
		this.fileToBeRead=file;
		readerType=cls;
		if(!file.exists()) throw new FileNotFoundException();
		if(file.isDirectory()) {
			File[] fileList;
			if((fileList=file.listFiles())==null || fileList.length==0) throw new IOException("no groups found");
			this.readerList=new BufferedReader [fileList.length];
			for(int i=0; i<fileList.length; i++) {
				this.readerList[i]=new BufferedReader(new FileReader(fileList[i]));
			}
		}else {
			this.readerList=new BufferedReader[1];
			this.readerList[0]=new BufferedReader(new FileReader(file));
		}
		this.currentReader=this.readerList[0];
		this.lastLine=this.currentReader.readLine();
	}
	
	@Override
	public boolean hasNext() {
		return lastLine!=null;
	}

	@Override
	@SneakyThrows
	public T next() {
		T res=this.readerType.getDeclaredConstructor(String.class).newInstance(this.lastLine);
		this.lastLine=currentReader.readLine();
		while(this.lastLine==null && this.currentReaderIndex+1<this.readerList.length) {
			this.currentReader=this.readerList[++this.currentReaderIndex];
			this.lastLine=currentReader.readLine();
		}
		return res;
	}

	@Override
	public Iterator<T> iterator() {
		// TODO Auto-generated method stub
		return this;
	}
	
	@SneakyThrows
	public void close() {
		for(int i=0; i<this.readerList.length; i++) {
			this.readerList[i].close();
		}
	}
	
	@SneakyThrows
	public ArrayList<T> getArrayList(){
		ArrayList <T> res = new ArrayList<T>();
		Reader<T> reader = new Reader<T>(this);
		for(T element: reader) {
			res.add(element);
		}
		reader.close();
		return res;
	}
	
	@SneakyThrows
	public Collection<T> getCollection(){
		return (Collection <T>) getArrayList();
	}
}
