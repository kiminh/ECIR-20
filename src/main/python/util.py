def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

