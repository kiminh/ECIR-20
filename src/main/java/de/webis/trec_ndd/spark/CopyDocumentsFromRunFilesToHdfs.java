package de.webis.trec_ndd.spark;

import java.util.List;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import de.webis.trec_ndd.spark.SparkBuild8GrammIndex.DocumentSelectionStrategy;
import de.webis.trec_ndd.trec_collections.AnseriniCollectionReader;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.collection.ClueWeb09Collection.Document;
import lombok.Getter;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

public class CopyDocumentsFromRunFilesToHdfs implements SparkArguments {
	
	@Getter
	private final Namespace parsedArgs;
	
	private CopyDocumentsFromRunFilesToHdfs(String[] args) {
		this.parsedArgs = parseArguments(args);
	}
	
	@Override
	public void run() {
		DocumentSelectionStrategy selectionStrategy = documentSelection();
		AnseriniCollectionReader<Document> acr = new AnseriniCollectionReader<Document>(collection());
		List<String> segmentPaths = acr.segmentPaths();

		try (JavaSparkContext context = context()) {
			JavaRDD<CollectionDocument> documents = context.parallelize(segmentPaths)
				.flatMap(s -> selectionStrategy.getDocumentsInSegmentPath().apply(acr).apply(s));
			
			documents.saveAsTextFile(jobName());
		}
	}
	
	public static void main(String[] args) {
		new CopyDocumentsFromRunFilesToHdfs(args).run();
	}
	
	private Namespace parseArguments(String[] args) {
		ArgumentParser parser = ArgumentParsers.newFor("CopyDocumentsToHdfs")
			.build()
			.defaultHelp(true)
			.description("Copy all selected documents to hdfs for faster access.");

		addCollectionToArgparser(parser);
		addDocumentSelectionToArgparser(parser);

		return parser.parseArgsOrFail(args);
	}
	
	public String jobName() {
		return jobName(collectionName(), documentSelection());
	}
	
	public static String jobName(String collectionName, DocumentSelectionStrategy documentSelection) {
		if(DocumentSelectionStrategy.ALL.equals(documentSelection)) {
			throw new RuntimeException("It is too expensive to use document-selection: " + documentSelection.name());
		}
		
		return "trec-docs-in-" + documentSelection.name().toLowerCase() + "-for-" + collectionName.toLowerCase();
	}
}
