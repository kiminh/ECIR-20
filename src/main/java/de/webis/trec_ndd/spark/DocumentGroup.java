package de.webis.trec_ndd.spark;

import java.io.Serializable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.TreeSet;
import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;
import scala.Tuple2;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class DocumentGroup implements Serializable {
	
	public String hash;
	public ArrayList<String> ids = new ArrayList<String>();
	
	public DocumentGroup (String line) {
		DocumentGroup parsedGroup = createFromJson(line);
		this.hash = parsedGroup.hash;
		this.ids = parsedGroup.ids;
	}
	
	public DocumentGroup (Tuple2<String, Iterable<Tuple2<String, String>>> group) {
		this.hash=group._1;
		group._2.forEach(tuple -> this.ids.add(tuple._2));
	}
	
	@SneakyThrows
	public String toString() {
		return new ObjectMapper().writeValueAsString(this);
	}
	
	@SneakyThrows
	public static List<DocumentGroup> readFromJonLines(Path path) {
		return Files.readAllLines(path).stream()
				.map(s -> createFromJson(s))
				.collect(Collectors.toList());
	}
	
	@SneakyThrows
	public static DocumentGroup createFromJson(String jsonLine) {
		return new ObjectMapper().readValue(jsonLine, DocumentGroup.class);
	}

	public boolean isUsed(HashSet<String> usedDocIDs) {
		for(String id : ids) if(usedDocIDs.contains(id)) return true;
		return false;
	}
	
	public DocumentGroup copy() {
		return new DocumentGroup(this.getHash(), new ArrayList<>(this.getIds()));
	}

	public static List<DocumentGroup> appendRetrievalEquivalentToContentEquivalent(List<DocumentGroup> contentEquivalent, List<DocumentGroup> retrievalEquivalent) {
		List<DocumentGroup> ret = new LinkedList<>(contentEquivalent);
		
		for(DocumentGroup retEq: retrievalEquivalent) {
			boolean shouldAddCompleteRetEq = Boolean.TRUE;
			
			for(DocumentGroup contEq: contentEquivalent) {
				for(String id: retEq.getIds()) {
					if(contEq.getIds().contains(id)) {
						contEq.getIds().addAll(retEq.getIds());
						shouldAddCompleteRetEq = Boolean.FALSE;
						continue;
					}
				}
			}
			
			if(shouldAddCompleteRetEq) {
				ret.add(retEq);
			}
		}
		
		for(DocumentGroup docGroup : ret) {
			docGroup.setIds(new ArrayList<>(new TreeSet<>(docGroup.getIds())));
		}
		
		return ret;
	}
}



