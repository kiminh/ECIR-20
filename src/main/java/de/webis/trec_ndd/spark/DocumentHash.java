package de.webis.trec_ndd.spark;

import java.io.Serializable;
import java.util.HashSet;
import java.util.List;

import org.codehaus.jackson.map.ObjectMapper;

import de.webis.trec_ndd.similarity.MD5;
import de.webis.trec_ndd.similarity.TextProfileSignatureSimilarity;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import de.webis.trec_ndd.util.NGramms.Word8Gramm;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class DocumentHash implements Serializable {
	private String id;
	private String textProfileSignature;
	private String md5;
	private String fullyTextProfileSignature;
	private String fullyCanonicalizedMd5;
	private long documentLength;
	private long fullyCanonicalizedWord8GrammCount;
	private long fullyCanonicalizedWord8GrammSetSize;
	
	public DocumentHash (CollectionDocument doc) {
		this.id=doc.getId();
		this.md5=MD5.md5hash(doc.getContent());
		this.textProfileSignature=TextProfileSignatureSimilarity.textProfileSignatureString(doc.getContent());
		this.documentLength=doc.getContent().length();
		this.fullyTextProfileSignature = TextProfileSignatureSimilarity.textProfileSignatureString(doc.getFullyCanonicalizedContent());
		this.fullyCanonicalizedMd5 = MD5.md5hash(doc.getFullyCanonicalizedContent());
		List<Word8Gramm> word8Gramms = NGramms.build8Gramms(doc.getFullyCanonicalizedContent());
		
		this.fullyCanonicalizedWord8GrammCount = word8Gramms.size();
		this.fullyCanonicalizedWord8GrammSetSize = new HashSet<>(word8Gramms).size();
	}
	
	@SneakyThrows
	public String toString() {
		return new ObjectMapper().writeValueAsString(this);
	}
	
	@SneakyThrows
	public static DocumentHash fromString(String str) {
		return new ObjectMapper().readValue(str, DocumentHash.class);
	}
}
