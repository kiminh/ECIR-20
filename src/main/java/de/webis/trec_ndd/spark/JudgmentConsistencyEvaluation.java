package de.webis.trec_ndd.spark;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.spark.api.java.JavaSparkContext;
import org.codehaus.jackson.map.ObjectMapper;

import com.google.common.collect.ImmutableMap;

import static de.webis.trec_ndd.spark.SparkDeduplicateAndEvaluate.readGroupsUsedInRunFiles;

import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.SharedTask.DocumentJudgments;
import lombok.Builder;
import lombok.Data;
import lombok.Getter;
import lombok.SneakyThrows;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

public class JudgmentConsistencyEvaluation implements SparkArguments {
	
	@Getter
	private final Namespace parsedArgs;
	
	public JudgmentConsistencyEvaluation(String[] args) {
		this.parsedArgs = parseArguments(args);
	}
	
	@Override
	public void run() {
		try (JavaSparkContext sc = context()) {
			for(TrecCollections collection: collections()) {
				String retrievalEquivalentDir = SparkGroupByFingerprint.resultDir(collection, documentSelection(), documentSimilarity());
				List<DocumentGroup> retrievalEquivalent = readGroupsUsedInRunFiles(sc, retrievalEquivalentDir, collection); 
				String contentEquivalentDir = SparkGroupByFingerprint.resultDir(collection, documentSelection(), "s3-similarity-connected-component-" + s3Threshold());
				List<DocumentGroup> contentEquivalent = readGroupsUsedInRunFiles(sc, contentEquivalentDir, collection);
				List<DocumentGroup> contentEquivalentWithoutRetrievalEquivalentOnly = new ArrayList<>(contentEquivalent.stream().map(DocumentGroup::copy).collect(Collectors.toList()));
				contentEquivalent = DocumentGroup.appendRetrievalEquivalentToContentEquivalent(contentEquivalent, retrievalEquivalent);
				
				DocumentGroups documentGroups = DocumentGroups.builder()
						.contentEquivalent(contentEquivalent)
						.retrievalEquivalent(retrievalEquivalent)
						.contentEquivalentWithoutRetrievalEquivalent(contentEquivalentWithoutRetrievalEquivalentOnly)
						.build();
				
				for(SharedTask task : collection.getSharedTasks()) {
					report(documentGroups, task);
				}
			}
		}
	}
	
	public static void main(String[] args) {
		new JudgmentConsistencyEvaluation(args).run();
	}
	
	@SneakyThrows
	private static void report(DocumentGroups docGroups, SharedTask task) {
		DocumentJudgments docJudgments = task.documentJudgments();
		
		Map<String, Object> report = ImmutableMap.<String, Object>builder()
				.put("SharedTask", task.name())
				.put("GroupSize", docGroups.getContentEquivalent().size())
				.put("InconsistentGroups", docGroups.inconsistentGroups(docJudgments))
				.put("InconsistentGroups(WithoutUnlabeled)", docGroups.inconsistentGroupsWithoutUnlabeled(docJudgments))
				.put("InconsistentGroups(WithoutUnlabeledAndRetrievalEquivalent)", docGroups.inconsistentGroupsWithoutUnlabeledWithoutRetrievalEquivalent(docJudgments))
				.put("ConsistentGroups", docGroups.consistentGroups(docJudgments))
				.put("JudgmentCount", docJudgments.judgmentCount())
				.put("DocumentsToJudgeAgain", docGroups.documentsToJudgeAgain(docJudgments))
				.put("TopicDetails", docGroups.topicDetails(docJudgments))
				.build();
		
		System.out.println(new ObjectMapper().writeValueAsString(report));
	}

	private Namespace parseArguments(String[] args) {
		ArgumentParser parser = ArgumentParsers.newFor("JudgmentConsistencyEvaluation")
				.build()
				.defaultHelp(true)
				.description("Evaluate inconsistencies caused by near-duplicates in the QRELs.");
		
		addCollectionsToArgparser(parser);
		addSimilarityToArgparser(parser);
		addDocumentSelectionToArgparser(parser);
		addS3Threshold(parser);
		
		return parser.parseArgsOrFail(args);
	}
	
	@Data
	@Builder
	public static class DocumentGroups {
		private List<DocumentGroup> contentEquivalent,
									retrievalEquivalent,
									contentEquivalentWithoutRetrievalEquivalent;
		
		public int inconsistentGroups(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistency(group) ? 1 : 0)
				.sum();
		}
		
		public int inconsistentGroupsWithoutUnlabeled(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistencyWithoutUnlabeled(group) ? 1 : 0)
				.sum();
		}
		
		public Object inconsistentGroupsWithoutUnlabeledWithoutRetrievalEquivalent(DocumentJudgments docJudgments) {
			return contentEquivalentWithoutRetrievalEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistencyWithoutUnlabeled(group) ? 1 : 0)
				.sum();
		}

		public int consistentGroups(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistency(group) ? 0 : 1)
				.sum();
		}
		
		public int documentsToJudgeAgain(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(docJudgments::documentsToJudgeAgain)
				.sum();
		}

		public Map<String, Map<String, Integer>> topicDetails(DocumentJudgments docJudgments) {
			return docJudgments.topics().stream()
				.collect(Collectors.toMap(i -> i, i -> topicDetails(i, docJudgments)));
		}
		
		private Map<String, Integer> topicDetails(String topic, DocumentJudgments docJudgments) {
			Set<String> relevantJudgments = new HashSet<>(docJudgments.getRelevantDocuments(topic));
			Set<String> irrelevantJudgments = new HashSet<>(docJudgments.getIrrelevantDocuments(topic));
			Set<String> retrievalEquivalent = getRetrievalEquivalent().stream()
				.flatMap(i -> i.ids.stream())
				.collect(Collectors.toSet());
			Set<String> contentEquivalent = getContentEquivalent().stream()
					.flatMap(i -> i.ids.stream())
					.collect(Collectors.toSet());
			
			failOnDocumentsThatAreRetrievalEquivalentButNotContentEquivalent(retrievalEquivalent, contentEquivalent, topic);
			
			return ImmutableMap.<String, Integer>builder()
				.put("relevantJudgments", relevantJudgments.size())
				.put("retrievalEquivalentGroupsInRelevantJudgments", groupsInIntersection(relevantJudgments, getRetrievalEquivalent()))
				.put("contentEquivalentGroupsInRelevantJudgments", groupsInIntersection(relevantJudgments, getContentEquivalent()))
				.put("retrievalEquivalentDocsInRelevantJudgments", intersection(retrievalEquivalent, relevantJudgments))
				.put("contentEquivalentDocsInRelevantJudgments", intersection(contentEquivalent, relevantJudgments))
				.put("irrelevantJudgments", irrelevantJudgments.size())
				.put("retrievalEquivalentDocsInIrrelevantJudgments", intersection(retrievalEquivalent, irrelevantJudgments))
				.put("contentEquivalentDocsInIrrelevantJudgments", intersection(contentEquivalent, irrelevantJudgments))
				.put("retrievalEquivalentGroupsInIrrelevantJudgments", groupsInIntersection(irrelevantJudgments, getRetrievalEquivalent()))
				.put("contentEquivalentGroupsInIrrelevantJudgments", groupsInIntersection(irrelevantJudgments, getContentEquivalent()))
				.build();
		}
		
		private static void failOnDocumentsThatAreRetrievalEquivalentButNotContentEquivalent(Set<String> retrievalEquivalent, Set<String> contentEquivalent, String topic) {
			List<String> notContentEquivalent = new ArrayList<>();
			
			for(String retrievalEq : retrievalEquivalent) {
				if(!contentEquivalent.contains(retrievalEq)) {
					notContentEquivalent.add(retrievalEq);
				}
			}
			
			if(notContentEquivalent.size() > 0) {
				throw new RuntimeException("Unexpected Behaviour for topic '" + topic + "': Documents are retrieval-equivalent but not content-equivalent: "+ notContentEquivalent);
			}
		}
		
		private int groupsInIntersection(Set<String> a, List<DocumentGroup> b) {
			int ret = 0;
			
			for(DocumentGroup docGroup: b) {
				if(docGroup.ids.stream().anyMatch(i -> a.contains(i))) {
					ret += 1;
				}
			}
			
			return ret;
		}
		
		private <T> int intersection(Set<T> a, Set<T> b) {
			Set<T> ret = new HashSet<T>(a);
			ret.retainAll(b);
			
			return ret.size();
		}
	}

	@Override
	public String jobName() {
		return "trec-qrel-consistency";
	}
}
