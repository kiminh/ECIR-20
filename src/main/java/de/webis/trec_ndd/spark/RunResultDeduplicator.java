package de.webis.trec_ndd.spark;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;

import de.webis.trec_ndd.trec_collections.SharedTask;
import lombok.Data;
import lombok.SneakyThrows;

public interface RunResultDeduplicator {

	public static final RunResultDeduplicator 
		base = buildDeduplicator((a,b) -> new DeduplicationResult(a, Collections.emptyMap()), "base"),
		removeDuplicates = buildDeduplicator(Internals::deduplicateRunResults, "removeDuplicates"),
		duplicatesMarkedIrrelevant = buildDeduplicator(Internals::duplicatesMarkedIrrelevant, "duplicatesMarkedIrrelevant");
	
	public DeduplicationResult deduplicateRun(List<RunLine> runLines, Collection<DocumentGroup> docGroups);

	public String name();
	
	public static List<RunResultDeduplicator> all(SharedTask task) {
		return Arrays.asList(
			base,
			removeDuplicates,
			duplicatesMarkedIrrelevant
		);
	}
	public static class Internals{
		
		public static DeduplicationResult duplicatesMarkedIrrelevant(List<RunLine> runLines, Collection<DocumentGroup> docGroups) {
			Map<Integer, Set<String>> topicToDocumentsToMarkAsIrrelevant = new HashMap<>();
			HashMap<RunLine, String> runLineMapGroup = helper(runLines, docGroups);
			HashSet<String> doneGroups = new HashSet<String>();
			ArrayList<RunLine> output = new ArrayList<RunLine>();
			int currentTopic=-1;
			for(RunLine rl: runLines) {
				if(currentTopic!=rl.getTopic()) {
					currentTopic=rl.getTopic();
				}
				if(!runLineMapGroup.containsKey(rl)) {
					output.add(rl.copy());
					continue;
				}
				
				String docGroupWithTopic = runLineMapGroup.get(rl)+rl.getTopic();
				
				if(!doneGroups.contains(docGroupWithTopic)){
					doneGroups.add(docGroupWithTopic);
					output.add(rl.copy());
				}
				else{
					if(!topicToDocumentsToMarkAsIrrelevant.containsKey(currentTopic)) {
						topicToDocumentsToMarkAsIrrelevant.put(currentTopic, new HashSet<>());
					}
					topicToDocumentsToMarkAsIrrelevant.get(currentTopic).add(rl.getDoucmentID());
					output.add(rl.copy());
				}
			}
			
			return new DeduplicationResult(output, topicToDocumentsToMarkAsIrrelevant); 
		}		
		
		private static HashMap<RunLine, String> helper(List<RunLine> runLines, Collection<DocumentGroup> docGroups) {
			
			// Read RunLine and map runline.id to runline
			HashMap<String,ArrayList<RunLine>> idMapRunLines = new HashMap<String,ArrayList<RunLine>>();
			for(RunLine rl : runLines) {
				if(!idMapRunLines.containsKey(rl.getDoucmentID())) idMapRunLines.put(rl.getDoucmentID(),new ArrayList<RunLine>());
				idMapRunLines.get(rl.getDoucmentID()).add(rl);
			}
			//Map RunLine to Group
			HashMap<RunLine, String> runLineMapGroup = new HashMap<RunLine,String>();
			for(DocumentGroup currentGroup: docGroups)
				for(String id: currentGroup.ids) {
					if(!idMapRunLines.containsKey(id)) continue;
					for(RunLine rl: idMapRunLines.get(id)) 
						runLineMapGroup.put(rl, currentGroup.hash);
				}
			return runLineMapGroup;
		}
		
		@SneakyThrows
		public static DeduplicationResult deduplicateRunResults(List<RunLine> runLines, Collection<DocumentGroup> docGroups) {
			Map<Integer, Set<String>> topicToDocumentsToMarkAsIrrelevant = new HashMap<>();
			HashMap<RunLine, String> runLineMapGroup = helper(runLines, docGroups);
			//copy runfile but lines with the same group, when a line of the group is already written
			HashSet<String> doneGroups = new HashSet<String>();
			int currentTopic=-1;
			int rankDiff=0;
			ArrayList<RunLine> output = new ArrayList<RunLine>();
			for(RunLine rl: runLines) {
				if(currentTopic!=rl.getTopic()) {
					currentTopic=rl.getTopic();
					rankDiff=0;
				}
				if(runLineMapGroup.containsKey(rl)) {
					if(!doneGroups.contains(runLineMapGroup.get(rl)+rl.getTopic())){
						doneGroups.add(runLineMapGroup.get(rl)+rl.getTopic());
					}
					else{
						rankDiff++;
						if(!topicToDocumentsToMarkAsIrrelevant.containsKey(currentTopic)) {
							topicToDocumentsToMarkAsIrrelevant.put(currentTopic, new HashSet<>());
						}
						topicToDocumentsToMarkAsIrrelevant.get(currentTopic).add(rl.getDoucmentID());
						
						continue; 
					}
				}
				output.add(rl.createNewWithRankMinus(rankDiff));
			}
			
			return new DeduplicationResult(output, topicToDocumentsToMarkAsIrrelevant); 
		}
	}
	
	@Data
	public static class DeduplicationResult {
		private final List<RunLine> deduplicatedRun;
		private final Map<Integer, Set<String>> topicToDocumentIdsToMakeIrrelevant;
	}

	public static RunResultDeduplicator buildDeduplicator(BiFunction<List<RunLine>, Collection<DocumentGroup>, DeduplicationResult> strategy, String name) {
		return new RunResultDeduplicator() {
			@Override
			public DeduplicationResult deduplicateRun(List<RunLine> runLines, Collection<DocumentGroup> docGroups) {
				return strategy.apply(runLines, docGroups);
			}

			@Override
			public String name() {
				return name;
			}
		};
	}
}
