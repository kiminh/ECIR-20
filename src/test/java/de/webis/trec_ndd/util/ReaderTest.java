package de.webis.trec_ndd.util;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;

import com.google.common.io.Files;

import de.webis.trec_ndd.spark.DocumentGroup;
import de.webis.trec_ndd.util.Reader;
import lombok.SneakyThrows;

public class ReaderTest {
	
	@Test
	@SneakyThrows
	public void NonExistingFile() {
		try {
			new Reader<DocumentGroup>("./src/test/ressources/DocumentGroupReader/nonexisting",DocumentGroup.class);
			Assert.assertTrue(false);
		}catch(Exception e) {
			if(e instanceof FileNotFoundException) Assert.assertTrue(true);
			else throw e;
		}
	}
	
	@Test
	@SneakyThrows
	public void File() {
		Reader<DocumentGroup> dgr = new Reader<DocumentGroup>("./src/test/resources/DocumentGroupReader/OneFile/group0",DocumentGroup.class);
		List<DocumentGroup> list =new ArrayList<DocumentGroup>();
		while(dgr.hasNext()) {
			list.add(dgr.next());
		}
		Approvals.verifyAsJson(list);
	}
	
	@Test
	@SneakyThrows
	public void EmptyFolder() {
		try {
			java.io.File emptyFolder = Files.createTempDir(); 
			new Reader<DocumentGroup>(emptyFolder.getAbsolutePath(),DocumentGroup.class);
			Assert.assertTrue(false);
		}catch(Exception e) {
			if(e instanceof IOException && e.getMessage()!=null && e.getMessage().equals("no groups found")) Assert.assertTrue(true);
			else throw e;
		}
	}
	
	@Test
	@SneakyThrows
	public void FolderWithOneFile() {
		Reader<DocumentGroup> dgr = new Reader<DocumentGroup>("./src/test/resources/DocumentGroupReader/OneFile",DocumentGroup.class);
		List<DocumentGroup> list =new ArrayList<DocumentGroup>();
		while(dgr.hasNext()) {
			list.add(dgr.next());
		}
		Approvals.verifyAsJson(list);
	}
	
	@Test
	@SneakyThrows
	public void FolderWithMoreThanOneFile() {
		Reader<DocumentGroup> dgr = new Reader<DocumentGroup>("./src/test/resources/DocumentGroupReader/MoreThanOneFile",DocumentGroup.class);
		List<DocumentGroup> list =new ArrayList<DocumentGroup>();
		while(dgr.hasNext()) {
			list.add(dgr.next());
		}
		list.sort(new Comparator<DocumentGroup>() {
		    public int compare(DocumentGroup obj1, DocumentGroup obj2) {
		        return obj1.getHash().compareTo(obj2.getHash());
		    }
		});
		Approvals.verifyAsJson(list);
	}

}
