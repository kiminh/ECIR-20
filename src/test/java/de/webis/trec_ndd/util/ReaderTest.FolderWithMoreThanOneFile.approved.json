[
  {
    "hash": "0cc470e47bf8e6bccf498887047c0086",
    "ids": [
      "clueweb09-ja0013-99-26143",
      "clueweb09-ja0015-22-20971"
    ]
  },
  {
    "hash": "19fc8dee9db49906028385e1d282cb76",
    "ids": [
      "clueweb09-en0003-30-19876",
      "clueweb09-en0013-15-07655"
    ]
  },
  {
    "hash": "1d16982e3511868d5da34d51df4719d8",
    "ids": [
      "clueweb09-en0079-57-38234",
      "clueweb09-en0023-21-22427"
    ]
  },
  {
    "hash": "1ecb8106b4974f0334d73c784f6756d4",
    "ids": [
      "clueweb09-en0058-75-23241",
      "clueweb09-en0100-26-06559"
    ]
  },
  {
    "hash": "21d416069c8618c2a82c652e3cd8cc52",
    "ids": [
      "clueweb09-en0005-67-29026",
      "clueweb09-en0005-67-29034"
    ]
  },
  {
    "hash": "29296ddb5d08db59b4f56f687576ca3b",
    "ids": [
      "clueweb09-en0047-51-00095",
      "clueweb09-en0026-66-03010",
      "clueweb09-en0011-82-24424"
    ]
  },
  {
    "hash": "2e9bb07bd758405157e96f6e9c2bc950",
    "ids": [
      "clueweb09-ko0002-97-23496",
      "clueweb09-ko0002-97-23497",
      "clueweb09-ko0002-97-23498",
      "clueweb09-ko0002-97-23499",
      "clueweb09-ko0002-97-23500",
      "clueweb09-ko0002-97-23501",
      "clueweb09-ko0002-97-23502",
      "clueweb09-ko0002-97-23503",
      "clueweb09-ko0002-97-23504",
      "clueweb09-ko0002-97-23505",
      "clueweb09-ko0002-97-23506",
      "clueweb09-ko0002-97-23507",
      "clueweb09-ko0002-97-23508",
      "clueweb09-ko0002-97-23509",
      "clueweb09-ko0002-97-23510",
      "clueweb09-ko0002-97-23511",
      "clueweb09-ko0002-97-23512",
      "clueweb09-ko0002-97-23513",
      "clueweb09-ko0002-97-23514",
      "clueweb09-ko0002-97-23515",
      "clueweb09-ko0002-97-23516",
      "clueweb09-ko0002-97-23517",
      "clueweb09-ko0002-97-23518"
    ]
  },
  {
    "hash": "35b9413ad7957d4b51f90afd3ab0c208",
    "ids": [
      "clueweb09-en0015-73-27081",
      "clueweb09-en0000-73-31679"
    ]
  },
  {
    "hash": "53065a145f3a5b7fa56f67639adecad5",
    "ids": [
      "clueweb09-en0009-71-27057",
      "clueweb09-en0088-56-31645"
    ]
  },
  {
    "hash": "5465bd43e50c6d271f1ba77388bae7f4",
    "ids": [
      "clueweb09-zh0009-35-18959",
      "clueweb09-zh0009-35-18962"
    ]
  },
  {
    "hash": "73613269a81412b40d7c4e018b8359e0",
    "ids": [
      "clueweb09-en0069-96-24946",
      "clueweb09-en0064-73-24946"
    ]
  },
  {
    "hash": "80630e116ba82a34d7516db35bfa8070",
    "ids": [
      "clueweb09-zh0036-08-11748",
      "clueweb09-zh0036-08-11760"
    ]
  },
  {
    "hash": "8334ecb572db9487b73b42f11d1dfd07",
    "ids": [
      "clueweb09-en0073-25-11125",
      "clueweb09-en0016-26-33209",
      "clueweb09-en0031-78-00636"
    ]
  },
  {
    "hash": "9bb37ce0c46347b04e7be3539db47ba1",
    "ids": [
      "clueweb09-zh0031-98-40771",
      "clueweb09-zh0019-22-42367"
    ]
  },
  {
    "hash": "a76e2d9e9c56b108fb1d6e15b2a73c2e",
    "ids": [
      "clueweb09-en0054-12-19351",
      "clueweb09-en0075-76-15478"
    ]
  },
  {
    "hash": "ac2d11c5d7f77cdef38f1566f3df8b05",
    "ids": [
      "clueweb09-ja0004-63-23900",
      "clueweb09-ja0009-69-49575",
      "clueweb09-ja0014-01-20176"
    ]
  },
  {
    "hash": "c1bae43910d8c3cf6c179de99e53dae0",
    "ids": [
      "clueweb09-en0129-24-20094",
      "clueweb09-en0117-59-28507"
    ]
  },
  {
    "hash": "d2c82a8d843d11401dace58ed47274cf",
    "ids": [
      "clueweb09-en0083-53-03677",
      "clueweb09-en0128-65-27891"
    ]
  },
  {
    "hash": "e6bd1b6c76128f172b6c3eb1958b3e67",
    "ids": [
      "clueweb09-de0004-28-20810",
      "clueweb09-de0004-28-20811"
    ]
  },
  {
    "hash": "fb31dc7716defb84cee553f5fb88c485",
    "ids": [
      "clueweb09-ja0006-21-30062",
      "clueweb09-ja0013-17-36554"
    ]
  }
]